### DISCUSSION TITLE
How should pyscaffold support the development process

### DISCUSSION THREAD

**Discussion by julie777** [pyscaffold/pyscaffold#601] [Dev Discussions]
Developing a software product is a process with many steps, many tools, and many variables. Python has always been a great language, but has not given developers full support the way that other languages have. Just the packaging step has always been difficult for new developers and only now after 30 years is starting to get easier.

This post ties in with my other discussions #592 and #598

I'll start of with a simple list of the various parts of the development lifecycle and then break it down some. Don't consider this complete or correct, because for now I am just trying to address the concept. I will start with some of the details from [software development](https://www.atlassian.com/software-development) and we can see what part of all this pyscaffold should be part of.

1. Concept
Projects are envisioned, designed, and prioritized. 
    Some of this information needs to be stored and for most OSS projects this is in github discussions, github wiki. But in very simple form just writing a README could be all it takes.

2. Planning/Roadmap
Stakeholders are identified, budgets set, and infrastructure requisitioned. Design documents from the concept phase are broken down into actionable tasks. 
    For this discussion let's just assume that github issues suffice.

3. Code/Review/Test
Development teams work to build production ready software that meets requirements and feedback. CI/CD pipelines ensure efficient developer experience.
    This is the part most developers gravitate towards and along with the next section where most tools that we are discussing fit.

4. Deploy/Release and Hosting
With code approved and merged, it’s time to ship it. 
   For our purposes let's just think of this as publishing to github for the source, pypi for the product, and readthedocs.io for the documentation.

5. Support
Support and maintenance is required of active software projects.
    For our purposes we can just think of this as being handled by GitHub issues.

6. Incident Management
Deprecation and end-of-life activities, including customer notification and migration are software development best practices.
    For our purposes we can just think of this as being handled by GitHub issues.

**Comment by julie777** [comment]
To what parts of the lifecycle should pyscaffold contribute?

I'm going just create a certainly incomplete list of lifecycle parts
1. create a github project
2. Create a development environment for a project on a developer's machine
    - create directory structure
    - create initial set of files
    - install tools
    - create virtualenv for isolated development
    - install dependencies
 
3. track project dependencies
4. build the project
5. run tests
6. create virtual environments and run tests in multiple environments
7. package the product
8. commit changes to repository (all project artifacts)
9. run continuous integration tests on changes uploaded to master repo
10. notify developers of failures
11. edit code
12. write tests
13. format code
14. run static checks on code
15. update issues
16. update docs
17. plan features
18. manage milestones
19. manage tool configuration
20. enforce standards


For a start :-)

**Reply by julie777** [reply]
This list isn't supposed to be complete, ordered or even correct, but rather just to give the discussion some context.

**Comment by abravalheri** [comment]
To me the real focus is item 2.

Of course, by including a `tox` template, we also help users to find default answers for several items of this list. However I don't think we should try to do everything...

What I like about the `tox` is that the provided template simplify the development for users, but allow them to completely customise which tools they want to use. We don't overtake the space of other tools, instead we just integrate them. (Like the UNIX philosophy).

**Reply by julie777** [reply]
And I understand that what pyscaffold does is 2 and in addition make it a little easier to have the tools for the others available. I just want to consider the possibility that pyscaffold could provide a simple interface for the novice user that allows doing everything, until more control is needed. I am still only expecting pysk to provide the actual tool functionality for 2.

**Comment by julie777** [comment]
@abravalheri 
> We don't overtake the space of other tools, instead we just integrate them. (Like the UNIX philosophy).

This is actually what I am trying to say, but I am looking at it just a little differently. Currently we create a project, and install some tools. The user has to know about the tools we install and use them. We rely on tox for many of the life cycle steps (build, multi-env testing, packaging, publishing to pypi), but there are many other steps that the developer need to take too. How does the developer figure out all the steps? How about learning all the tools (dependency management, writing tests, running tests, etc.)?

I know I have mentioned Hatch too much, but it provides a real benefit to the new user. `--help` provides a list of commands and those commands provide a guide to the steps in the development lifecycle that the user needs to follow. At the novice level the actual tools for each part of the lifecycle is irrelevant because the scaffold provides the means to run the necessary commands. Later and for more advanced use the user can get to know each tool.

Here are a couple quick (and ugly) sketches with just enough detail to get the idea across.

Now

![image](https://user-images.githubusercontent.com/5090005/153963951-4f8645c4-ca38-47d8-aafa-cc933abc8c2f.png)

Possibly

![image](https://user-images.githubusercontent.com/5090005/153964000-4d347614-4ca8-4954-9676-725cd580077d.png)

Which gives us the options of make the functionality easily accessible to the user and the options to swap tools for each function part as new tools and techniques are developed. Tool and Functions provided don't have to be 1 to 1.

**Comment by julie777** [comment]
While trying to learn, I have been working on a drawing that captures all the bits of functionality that pysk implements, what tools implement, how extensions connect, etc.  The drawing isn't at all shareable, but a partial list of what pysk currently implements is:

- cli
- api
- logging
- variable processing (configuration, commandline, environment, extension, template)
- creating project structure (from definitions in code)
- creating project files (from templates shipped with pysk)
- template expansion
- action interpreter
    - builtin actions
    - "builtin" commands
- tool connection points
- extension connection

And it provides data in the form of
- structure definition
- file templates
- default configuration for tools
- some default dependencies

I realize that this is the parts and not the details of how they are connected. You guys have years of experience with how they connect and how one tool might interact (badly?) with another. I'll be trying to catch up for quite a while, so if I write something that makes no sense just tell me, and please tell me why.  (I am having fun working with you!)

**Comment by julie777** [comment]
@abravalheri 
I copied your post from the issue because we are not discussing the issue specifically any longer.
> I am afraid I am still failing to see how that vision is incompatible with tox (or nox) 😅
>
> I think all the arguments you gave are even more reasons to use tox.
>
> Tox is not really a test runner... It is this quirky task runner tool unique to the Python system that is able to run general commands.
>
> It is not really comparable to make, ninja or just because it does not handle task dependencies... But on the other hand it will automatically bootstrap virtual environments which is super handy in a Python-centric dev environment.
>
> If you want to use flit, twine, poetry, mypy, sphinx, mkdocs... it is completely fine, you can do that with tox.

I spent a couple hours reading the link your provided for tox, reading [contributing](https://pyscaffold.org/en/v4.1.3_a/contributing.html), reading tox documentation, and experimenting with tox I have a much better understanding of your point of view. 

I haven't been using Python for the last couple years, so I am not up-to-date on the current status of everything in the ecosystem yet.

At my last job, building SlingTV, we completely controlled all aspects of the deployed environment, so we only had to support on version of python and really didn't need much of what tox offers. Our CI was using Jenkins connected to our own instance of a github like repo.

### What have I learned?

- After all the time spent learning more about tox I definitely see where your views come from. 
- I can see how PyScaffold was modeled after tox in many ways. (Invoking actions, one command with many options, etc.)
- I can even see that tox could run putup, like it does for setuptools, flit, or other tools, and then the user could do everything with tox.
- I completely understand that my view of tox as a test runner for multiple environments is seeing only a small part of the picture.

Mimicking a successful product is usually a good thing. I'm not sure that is true in this case. As was mentioned, tox started a very good multi-env test runner. That is reflected in its use of a single command "tox".

I remember a mention that "tox uses [testenv] for historical reasons even though it is confusing" or something similar.  This results in a non-obvious way of doing things for the tox user. I certainly don't think it is a good way for a new Python developer to work. I struggled for a bit to understand that there are no tox commands. I can't tell tox to build my project. I can't tell tox to test my project. All I can do is tell tox to run using [env].

This morning I created a new project to do some exploring. I did follow the steps in contributing to enable running from dev code. 
I expected to be able to run the example tests.

When I typed `pytest`, I found that it wasn't installed...
When I ran `pip list` I found that the dev tools were not installed.

When I ran `tox` it created virtual environments and ran tests in those environments. Extrapolating this to a real project where the [write test code, run test, write code, run test] cycle should be extremely fast, I don't want to run tox. I don't want to run coverage or all the other informational parts. I really want just run pytest.

I did some reading and then added
```[testenv:dev]
description = environment for coding and running tests
commands =
deps =
```
to tox.ini and now I can run `tox` to run pytest in my dev environment.

I think this little peek into what I (in the role of a new pysk user) would experience out of the box shows why I think we can provide a better experience. From my perspective, pysk expects users to be experienced tox users.

This also clarifies some of my earlier comments about pysk and what running putup does. 

- I expected it to create a project that was ready to push a button and publish to pypi, although with no functionality.
- Then I tried it and I thought that it was creating a project without doing any of the setup, such as creating a venv, installing dependencies, etc.
- Now I realize that it is creating a project that is ready to publish to pypi, but you just have to use tox to do it, and use tox for other steps of the dev process.

Now with my new undestanding of tox it appears that pysk supports the following commands for the developer with a default project
- putup
- git ...
- tox -e build
- tox -e clean
- tox    (implicit -e test)
- tox -e docs
- tox -e doctests
- tox -e linkcheck
- tox -e publish

Some dependency and tool config is in the setup.cfg
backend (tool) dependencies are in tox.ini
some dependencies are in pyproject.toml
package dependencies go in setup.cfg

Because tox runs all the backend tools, tox.ini has their dependencies, and I assume that they won't be managed and have their versions frozen by whatever tools is used for package dependencies.

And I have to conclude that I fell into the trap of most users and didn't read the docs. I skimmed them and then just tried to start working. Much of what I now know is in the docs. You can see from what I tried above that I wasn't reading the quick start doc. ☹️ 

I will start back at the beginning and read all the documentation before my next post.

**Comment by FlorianWilhelm** [comment]
I have actually never heard of [Hatch](https://ofek.dev/hatch/) before you mentioned it @julie777. Looking at its docs and I just tested it a bit, it seems to do a few more things compared to PyScaffold. It also seems to integrate the management of virtualenv. In doing so it becomes quite similar to [Poetry](https://python-poetry.org/). A difference between these two is that Poetry uses their home-grown build backend while Hatch uses setuptools. 

So let's try to sum up the differences between Hatch und PyScaffold.

 | Functionality | Hatch | PyScaffold |
|---------------|-------|------------|
| Unit tests | pytest | pytest |
|   Test runner             |  [home-grown](https://ofek.dev/hatch/latest/environment/#matrix)      |       tox     |
|   Virtualenv Mgmt            |  [home-grown](https://ofek.dev/hatch/latest/environment/) based on virtualenv      |  no supported  but documented        |
|    Versioning           | [home-grown](https://ofek.dev/hatch/latest/version/) but manual      |     setuptools-scm automatically takes git tags      |
| Publishing | [home-grown](https://ofek.dev/hatch/latest/publish/) | twine |
| Building  | [home-grown interface](https://ofek.dev/hatch/latest/build/), internally setuptools |  tox or whatever works with setuptools|
| Docs | unsupported | Sphinx |
| Misc. | coverage | coverage, pre-commit, CI/CD, RTD, flake8, ... |

So Hatch tries to unify everything and thus warps things like virtualenv management. PyScaffold on the other hand gives you a fine selection of the tools you would want to use and also sane defaults for them but doesn't wrap them. There are two reasons for this, first its the unix philosophy (that I quite like) and it's so much easier to maintain (a lazy programmer is a good programmer, and I am definitely lazy ;-)  

So at least for me, managing virtualenv is out of scope for PyScaffold. There are people that like pipenv, pip-tools, plain virtualenv, conda/mamba, pyenv, ... (there is so many) and thus we leave it to extension developers to integrate them and have basic support in the core only for virtualenv with `--venv` and document the integration with pip-tools, pipenv and conda. Besides the fact that there are many competing tools for venv management, there is also the fact that both tasks, i.e. venv mgmt & project setup, are orthogonal to eachother. So why mix them in one tool?

**Reply by julie777** [reply]
@FlorianWilhelm  I like your table. It is a much more focused version of the one I have in my notes that is many pages long.

I have been reading, and reading and cogitating on this.

We have to have some about of integration for quite a few tools, either because we actually use them, or we set them up (configure) them for the user. It seems like defining the function of the tool, such as publishing and how one could provide all the information required for us to use the tool would be a good idea. By this I mean really being clear about functional steps like packaging or publishing.

I did the refactoring to consolidate all the repo(git) related code. I still need to finish testing and submit a pull request. But I have been thinking about doing an experiment and separating out the repo requirements/api from the git implementation of that. That would allow for a possible future where we had support for mercurial also.

This led me to consider users that for some reason wanted or were required to use tools other than our defaults. It is awkward to use extensions and dozens of commandline arguments for this. The config file is a better place for the user to describe what they want all the time.

That led me to wonder about config files and whether is a difference between the pyscaffold config and the config we save from commandline options for a project. To give you an example
- pyscaffold config would contain the list of tools I want to use
- pysk-app.cfg would contain the configuration for creating an app project
- pysk-package would contain the configuration for creating a package project

I realize that this goes against what you have been thinking, but if we really want to be significant in the Python ecosystem we need to capture a large number of users. I looked at the project stats from PyPI and we really aren't making the kind of impact I would like to see.

I realize that this has gotten pretty long. I wanted to share my thought process and see if any of it makes sense to you guys.

I am going to shift to another discussion for thoughts about "do one thing and do it well". I'll add it to #592 since that isn't getting a response.

**Reply by FlorianWilhelm** [reply]
Please also read my response [here](https://github.com/pyscaffold/pyscaffold/discussions/608#discussioncomment-2213453) as it is relevant for it.

**Reply by abravalheri** [reply]
Hi @julie777,

> * pysk-app.cfg would contain the configuration for creating an app project
> * pysk-package would contain the configuration for creating a package project

Currently it is possible to do that with:

```
putup myproj --config ~/.config/pyscaffold/app.cfg
putup myproj --config ~/.config/pyscaffold/lib.cfg
```

There is also the default `~/.config/pyscaffold/default.cfg` that is automatically considered by `putup myproj`.

**Comment by abravalheri** [comment]
Currently the `--venv` extension does not install the project being created automatically. Would that change help with the challenges?

The side effect of making the installation of the project implicit is that users might also expect that when they change the dependencies or entry-points, everything is magically updated... Which will not happen, they need to refresh the virtual environment by running `pip install` again.

---

I don't think that PyScaffold should be used to manage virtual environments directly. The only reason I added `--venv` was to automate an repetitive workflow process.

To me, using `putup myproj --venv --venv-install tool` is just a shorthand for:
```
putup myproj
virtualenv myproj/.venv
myproj/.venv/bin/pip install tool
```
(i.e. boilerplate tasks that I run all the time I start a new project).

**Comment by julie777** [comment]
> putup myproj
virtualenv myproj/.venv
myproj/.venv/bin/pip install tool

@abravalheri 

This is exactly my point. And I suspect that you do even more than this. You probably install multiple tools. You probably do a pip dev install so you can run the code you are working on.

Why wouldn't you want to get rid of typing boilerplate tasks and have them done automatically?
Please help me understand.

**Reply by abravalheri** [reply]
> Why wouldn't you want to get rid of typing boilerplate tasks and have them done automatically?
Please help me understand.

I am completely fine with that.
The only thing that I am afraid here is users thinking that the virtualenv should automatically update if they change dependencies in `setup.cfg` or if they add new entry-points etc...

**Comment by julie777** [comment]
@abravalheri  
> I am completely fine with that.
The only thing that I am afraid here is users thinking that the virtualenv should automatically update if they change dependencies in setup.cfg or if they add new entry-points etc...

I understand your concern. 
-  if you use pip to install additional dependencies in your venv it is updated but the dependencies haven't been recorded to use in a build
- If just just add them to setup.cfg then they are available for the build, but not installed in the venv

I have never used `setup.cfg` for dependencies. In the past I have always used requirements.txt and `pip freeze`, etc. so I am not really sure about the effects of dependencies in setup.cfg.

I am reading setuptools doc to make sure I understand setuptools better. Maybe you can help me a little.
If I put dependencies in setup.cfg, then to install them after do I run `pip install -e .` again?

I have a similar lack of understanding of putting all the dev dependencies in `tox.ini`. I believe that prevents using `pip` to install them in a dev venv.

And finally, I am rereading all the PEPs about pyproject.toml so that I can participate in discussions about it intelligently. For example, I am not sure if dependencies are ultimately  supposed to end up there. They can't be in setup.cfg when using a backend other than setuptools.

**Reply by FlorianWilhelm** [reply]
There are bascially 3 kinds of requirements @julie777:

* concrete requirements: Those go into `requirements.txt` or `environment.yml` and should be as specific as possible (maybe pinned) as they allow to reproduce what was done.
* abstract requirements: Those go into `setup.cfg` or `pyproject.toml` in the future. The idea is that every direct dependency (so what you directly use in the source) is listed and not pinned, maybe only like a lower version requirement. This is important if people use `pip` to install the requirement. If you for instance install `pandas` it will also pull `numpy`. No one gives you a requirement.txt and tells you how to set up an environment for them. This would in fact be dangerous because if two libs would use numpy and pin them to different version, they whole thing wouldn't work
* test requirements: those are used by tox to run the various test. Those might be dependencies you don't need in your actual dev environment and definitely not in the concrete requirements.
